src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -268,327 +343,172 @@ protected BoundExpression LowerTest(BoundDagTest test)
-                                    TryLowerTypeTestAndCast(testNode.Test, e.Evaluation, out BoundExpression sideEffect, out BoundExpression testExpression))
-                                {
-                                    _sideEffectBuilder.Add(sideEffect);
-                                    AddConjunct(testExpression);
-                                    node = e.Next;
-                                }
-                                else
-                                {
-                                    LowerOneTest(testNode.Test);
-                                    node = testNode.WhenTrue;
-                                }
-                            }
-                            break;
+                        // we've arranged to use the input value from the variable it is already stored in
+                    }
+                    else
+                    {
+                        var tempToHoldInput = _tempAllocator.GetTemp(temp);
+                        addCode(_factory.AssignmentExpression(tempToHoldInput, expr));
                     }
                 }
 
-                // When we get to "the end", we see if it is a success node.
-                switch (node)
+                BoundDecisionDagNode makeReplacement(BoundDecisionDagNode node, Func<BoundDecisionDagNode, BoundDecisionDagNode> replacement)
                 {
-                    case BoundLeafDecisionDagNode leafNode:
-                        {
-                            if (leafNode.Label == failureLabel)
-                            {
-                                // It is not clear that this can occur given the dag "optimizations" we performed earlier.
-                                AddConjunct(_factory.Literal(false));
-                            }
-                            else
-                            {
-                                Debug.Assert(leafNode.Label == successLabel);
-                            }
-                        }
-
-                        break;
-
-                    case BoundWhenDecisionDagNode whenNode:
-                        {
-                            Debug.Assert(whenNode.WhenExpression == null);
-                            Debug.Assert(whenNode.WhenTrue is BoundLeafDecisionDagNode d && d.Label == successLabel);
-                            foreach ((BoundExpression left, BoundDagTemp dagTemp) in whenNode.Bindings)
+                    switch (node)
+                    {
+                        case BoundEvaluationDecisionDagNode evalNode:
+                            if (evalNode.Evaluation is BoundDagFieldEvaluation eval &&
+                                eval.Input.IsOriginalInput &&
+                                eval.Field is var field &&
+                                field.IsTupleField &&
+                                field.CorrespondingTupleField != null &&
+                                field.TupleElementIndex is int i)
```

Comment:
Can this be simplified, by removing `field.IsTupleField` or `field.CorrespondingTupleField != null`?